File: src/main/java/com/hwscs/backend/service/interfaces/ShiftRequestService.java
package com.hwscs.backend.service.interfaces;

import com.hwscs.backend.dto.requests.CreateShiftRequestDto;
import com.hwscs.backend.dto.requests.InchargeApprovalDto;
import com.hwscs.backend.dto.response.EligiblePeerDto;
import com.hwscs.backend.dto.response.PeerResponseDto;
import com.hwscs.backend.dto.response.RequestHistoryResponseDto;
import com.hwscs.backend.dto.response.ShiftRequestResponseDto;

import java.util.List;

public interface ShiftRequestService {
    ShiftRequestResponseDto createRequest(CreateShiftRequestDto dto, String requesterUsername);

    ShiftRequestResponseDto peerResponse(PeerResponseDto dto, String peerUsername);

    ShiftRequestResponseDto inchargeReview(InchargeApprovalDto dto, String inchargeUsername);

    // Nurse: view all requests they are involved in
    List<ShiftRequestResponseDto> getMyRequests(String nurseUsername);

    // Nursing Incharge: view all PEER_ACCEPTED requests in their department
    List<ShiftRequestResponseDto> getPendingInchargeReview(String inchargeUsername);

    // View a single request by ID
    ShiftRequestResponseDto getRequestById(Integer requestId);

    // List all eligible peer nurses for shift
    List<EligiblePeerDto> getEligiblePeers(
            Integer requesterShiftId,
            String requesterUsername
    );

    // Cancel Request
    ShiftRequestResponseDto cancelRequest(Integer requestId, String requestUsername);

    // Get Request History
    List<RequestHistoryResponseDto> getRequestHistory(
            Integer requestId,
            String username
    );
}

File: frontend/src/app/core/services/shift-request.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { CreateShiftRequestDto, EligiblePeerDto, PeerResponseDto, RequestHistoryResponseDto, ShiftRequestResponseDto, InchargeApprovalDto } from '../models/models';

@Injectable({
  providedIn: 'root'
})
export class ShiftRequestService {
  private readonly apiUrl = '/api/shift-requests';

  constructor(private http: HttpClient) {}

  createRequest(dto: CreateShiftRequestDto): Observable<ShiftRequestResponseDto> {
    return this.http.post<ShiftRequestResponseDto>(this.apiUrl, dto);
  }

  peerResponse(dto: PeerResponseDto): Observable<ShiftRequestResponseDto> {
    return this.http.put<ShiftRequestResponseDto>(`${this.apiUrl}/peer-response`, dto);
  }

  getMyRequests(): Observable<ShiftRequestResponseDto[]> {
    return this.http.get<ShiftRequestResponseDto[]>(`${this.apiUrl}/my-requests`);
  }

  getEligiblePeers(requesterShiftId: number): Observable<EligiblePeerDto[]> {
    const params = new HttpParams().set('requesterShiftId', requesterShiftId.toString());
    return this.http.get<EligiblePeerDto[]>(`${this.apiUrl}/eligible-peers`, { params });
  }

  cancelRequest(id: number): Observable<ShiftRequestResponseDto> {
    return this.http.put<ShiftRequestResponseDto>(`${this.apiUrl}/${id}/cancel`, {});
  }

  getRequestHistory(id: number): Observable<RequestHistoryResponseDto[]> {
    return this.http.get<RequestHistoryResponseDto[]>(`${this.apiUrl}/${id}/history`);
  }

  getPendingReview(): Observable<ShiftRequestResponseDto[]> {
    return this.http.get<ShiftRequestResponseDto[]>(`${this.apiUrl}/pending-review`);
  }

  inchargeReview(dto: InchargeApprovalDto): Observable<ShiftRequestResponseDto> {
    return this.http.put<ShiftRequestResponseDto>(`${this.apiUrl}/incharge-review`, dto);
  }
}

